app/controllers/application_controller.rb

```diff
@@ -7,4 +7,131 @@ class ApplicationController < Sinatra::Base
     set :views, 'app/views'
   end
 
-end
\ No newline at end of file
+  configure do
+    enable :sessions
+    set :session_secret, "secret"
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+
+    if session[:id]
+      redirect to '/tweets'
+    end
+    erb :'/signup'
+  end
+
+  post '/signup' do
+    @user = User.new(username: params["username"], email: params["email"], password: params["password"])
+
+     if @user.save
+       session[:id] = @user.id
+       redirect to '/tweets'
+     else
+         redirect to '/signup'
+     end
+
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect to '/tweets'
+    end
+    erb :'login'
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:id] = @user.id
+      redirect to '/tweets'
+    end
+  end
+
+  get '/logout' do
+    session.clear
+    redirect to '/login'
+  end
+
+
+  get '/tweets' do  #loads home page after user is in
+    if logged_in?
+      @tweets = Tweet.all
+      erb :'tweets/index'
+    else
+      redirect to '/login'
+    end
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```